index/segment/mem/simple_terms_dict.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"regexp"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+)
+
+// simpleTermsDictionary uses two-level map to model a terms dictionary.
+// i.e. fieldName -> fieldValue -> postingOffset
+type simpleTermsDictionary struct {
+	fieldNamesLock sync.RWMutex
+	fieldNames     map[string]*fieldValuesMap
+
+	postingsList postingsList
+}
+
+func newSimpleTermsDictionary(initialSize int, fn newPostingsListFn) termsDictionary {
+	return &simpleTermsDictionary{
+		fieldNames:   make(map[string]*fieldValuesMap, initialSize),
+		postingsList: fn(),
+	}
+}
+
+func (t *simpleTermsDictionary) Insert(field doc.Field, i segment.DocID) error {
+	fieldName := string(field.Name)
+	fieldValues := t.getOrAddFieldName(fieldName)
+	fieldValue := string(field.Value)
+	return fieldValues.addDocIDForValue(fieldValue, i, t.postingsList)
+}
+
+func (t *simpleTermsDictionary) Fetch(f segment.Filter) (docIDsSet, error) {
```

Comment:
Should we support multiple filters here so we can do things like: `service:foo AND env:production`?

Reply:
`segment.Query()` supports the kind of conjunctions you're thinking of. The `Fetch()` on a terms dicts is per filter